File: algorithms/Path Sum II/Path Sum II.py
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res = []
        def dfs(v, path, pathsum):
            if not v:
                return
            path.append(v.val)
            pathsum += v.val
            if not v.left and not v.right and pathsum == targetSum:
                res.append(path[:])
            dfs(v.left, path, pathsum)
            dfs(v.right, path, pathsum)
            path.pop()
        dfs(root, [], 0)
        return res

File: scripts/Path Sum II/Path Sum II.java
class Solution 
{
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) 
    {
        List<List<Integer>> ans = new ArrayList<>();
        pathSum(root, targetSum, new ArrayList<>(), ans);
        return ans;
    }
    
    public void pathSum(TreeNode root, int targetSum, List<Integer> path, List<List<Integer>> ans)
    {
        if(root == null)
            return;
        path.add(root.val);
        if(root.left == null && root.right == null && targetSum == root.val)//leaf node that completes path
        {
            ans.add(new ArrayList(path));// we use new ArrayList because if we don't the originaly List is added which is mutable, if we add a copy that's not mutable.
        }
        else
        {
            pathSum(root.left, targetSum-root.val, path, ans);
            pathSum(root.right, targetSum-root.val, path, ans);
        }
        path.remove(path.size()-1); //removal of redundant nodes
    }
}
